src/ops.rs

```diff
@@ -707,21 +746,338 @@ pub fn imul3_32bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -
     Ok(())
 }
 
-pub fn shl_8bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+pub fn rol_8bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let mut destination = vm.get_arg(pipeline.args[0].location)?.u8_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u8_exact()?;
+    count = count & 0x1F;
+    count = count % 8;
+    if count == 0 {
+        return Ok(());
+    }
+    destination = (destination << count) | (destination >> (8 - count));
+    vm.flags.carry = (destination & 0x1) != 0;
+    if count == 1 {
+        vm.flags.overflow = destination.get::<BigEndian>(0.into()) as bool ^ vm.flags.carry;
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Byte(destination))?;
+    Ok(())
+}
+
+pub fn rol_native_word(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    if pipeline.size_override {
+        return rol_16bit(vm, pipeline, _hv);
+    } else {
+        return rol_32bit(vm, pipeline, _hv);
+    }
+}
+
+pub fn rol_16bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let mut destination = vm.get_arg(pipeline.args[0].location)?.u16_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u16_sx()?;
+    count = count & 0x1F;
+    count = count %16;
+    if count == 0 {
+        return Ok(());
+    }
+    destination = (destination << count) | (destination >> (16 - count));
+    vm.flags.carry = (destination & 0x1) != 0;
+    if count == 1 {
+        vm.flags.overflow = destination.get::<BigEndian>(0.into()) as bool ^ vm.flags.carry;
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Word(destination))?;
+    Ok(())
+}
+
+pub fn rol_32bit(vm: &mut VM, pipeline: &Pipeline, _hv: &mut dyn Hypervisor) -> Result<(), VMError>{
+    let mut destination = vm.get_arg(pipeline.args[0].location)?.u32_exact()?;
+    let mut count = vm.get_arg(pipeline.args[1].location)?.u32_sx()?;
+    count = count & 0x1F;
+    count = count %32;
+    if count == 0 {
+        return Ok(());
+    }
+    destination = (destination << count) | (destination >> (32 - count));
+    vm.flags.carry = (destination & 0x1) != 0;
+    if count == 1 {
+        vm.flags.overflow = destination.get::<BigEndian>(0.into()) as bool ^ vm.flags.carry;
+    }
+    vm.set_arg(pipeline.args[0].location, SizedValue::Dword(destination))?;
+    Ok(())
+}
+
+#[allow(exceeding_bitshifts)]
```

Comment:
Is this because Rust can't properly determine the range of `count` and thus throws the bitshift error for safety? From what I see, count should never exceed 8.

Reply:
Rust doesn't allow bitshifts to exceed the size of the integer by default and in this case we do need to be able to do that.

Reply:
But yes. It can't properly determine the range of `count` and `destination`